File: Packet/ServerUpdate.cs
using System.Collections.Generic;
using System.IO;
using Resources.Utilities;

namespace Resources.Packet {
    public class ServerUpdate : Packet {
        public class BlockDelta {
            public IntVector position;
            public ByteVector color;
            public BlockType type;
            public int unknown;

            public BlockDelta() { }
            public BlockDelta(BinaryReader reader) {
                position = new IntVector(reader);
                color = new ByteVector(reader);
                type = (BlockType)reader.ReadByte();
                unknown = reader.ReadInt32();
            }

            public void Write(BinaryWriter writer) {
                position.Write(writer);
                color.Write(writer);
                writer.Write((byte)type);
                writer.Write(unknown);
            }
        }
        public class Particle {
            public LongVector position;
            public FloatVector velocity;
            public FloatVector color;
            public float alpha;
            public float size;
            public int count;
            public ParticleType type;
            public float spread;
            public int unknown;

            public Particle() { }
            public Particle(BinaryReader reader) {
                position = new LongVector(reader);
                velocity = new FloatVector(reader);
                color = new FloatVector(reader);
                alpha = reader.ReadSingle();
                size = reader.ReadSingle();
                count = reader.ReadInt32();
                type = (ParticleType)reader.ReadInt32();
                spread = reader.ReadSingle();
                unknown = reader.ReadInt32();
            }

            public void Write(BinaryWriter writer) {
                position.Write(writer);
                velocity.Write(writer);
                color.Write(writer);
                writer.Write(alpha);
                writer.Write(size);
                writer.Write(count);
                writer.Write((int)type);
                writer.Write(spread);
                writer.Write(unknown);
            }
        }
        public class Sound {
            public IntVector position;
            public SoundID soundID;
            public float pitch;
            public float volume;

            public Sound() { }
            public Sound(BinaryReader reader) {
                position = new IntVector(reader);
                soundID = (SoundID)reader.ReadInt32();
                pitch = reader.ReadSingle();
                volume = reader.ReadSingle();
            }

            public void Write(BinaryWriter writer) {
                position.Write(writer);
                writer.Write((int)soundID);
                writer.Write(pitch);
                writer.Write(volume);
            }
        }
        public class StaticEntity {
            public int chunkX;
            public int chunkY;
            public int id;
            public int paddingA;
            public StaticEntityType type;
            public int paddingB;
            public LongVector position;
            public int rotation;
            public FloatVector size;
            public bool closed; //bool?
            //3pad
            public int time; //for open/close doors ect
            public int unknown;
            public int paddingC;
            public long guid; //of player who interacts with it

            public StaticEntity() { }
            public StaticEntity(BinaryReader reader) {
                chunkX = reader.ReadInt32();
                chunkY = reader.ReadInt32();
                id = reader.ReadInt32();
                paddingA = reader.ReadInt32();
                type = (StaticEntityType)reader.ReadInt32();
                paddingB = reader.ReadInt32();
                position = new LongVector(reader);
                rotation = reader.ReadInt32();
                size = new FloatVector(reader);
                closed = reader.ReadBoolean();
                reader.ReadBytes(3);
                time = reader.ReadInt32();
                unknown = reader.ReadInt32();
                paddingC = reader.ReadInt32();
                guid = reader.ReadInt64();
            }

            public void Write(BinaryWriter writer) {
                writer.Write(chunkX);
                writer.Write(chunkY);
                writer.Write(id);
                writer.Write(paddingA);
                writer.Write((int)type);
                writer.Write(paddingB);
                position.Write(writer);
                writer.Write(rotation);
                size.Write(writer);
                writer.Write(closed);
                writer.Write(new byte[3]);
                writer.Write(time);
                writer.Write(unknown);
                writer.Write(paddingC);
                writer.Write(guid);
            }
        }
        public class ChunkItems {
            public class DroppedItem {
                public Item item;
                public long posX;
                public long posY;
                public long posZ;
                public float rotation;
                public float scale;
                public int unknownA; //bool?
                public int droptime;
                public int unknownB;
                public int unknownC;

                public DroppedItem() { }
                public DroppedItem(BinaryReader reader) {
                    item = new Item(reader);
                    posX = reader.ReadInt64();
                    posY = reader.ReadInt64();
                    posZ = reader.ReadInt64();
                    rotation = reader.ReadSingle();
                    scale = reader.ReadSingle();
                    unknownA = reader.ReadInt32();
                    droptime = reader.ReadInt32();
                    unknownB = reader.ReadInt32();
                    unknownC = reader.ReadInt32();
                }

                public void Write(BinaryWriter writer) {
                    item.Write(writer);
                    writer.Write(posX);
                    writer.Write(posY);
                    writer.Write(posZ);
                    writer.Write(rotation);
                    writer.Write(scale);
                    writer.Write(unknownA);
                    writer.Write(droptime);
                    writer.Write(unknownB);
                    writer.Write(unknownC);
                }
            }

            public int chunkX;
            public int chunkY;
            public List<DroppedItem> droppedItems = new List<DroppedItem>();

            public ChunkItems() { }
            public ChunkItems(BinaryReader reader) {
                chunkX = reader.ReadInt32();
                chunkY = reader.ReadInt32();
                int m = reader.ReadInt32();
                for (int i = 0; i < m; i++) {
                    droppedItems.Add(new DroppedItem(reader));
                }
            }

            public void Write(BinaryWriter writer) {
                writer.Write(chunkX);
                writer.Write(chunkY);
                writer.Write(droppedItems.Count);
                foreach (DroppedItem droppedItem in droppedItems) {
                    droppedItem.Write(writer);
                }
            }
        }
        public class P48 {
            public class SubP48 {
                public ulong unknownA;
                public ulong unknownB;

                public SubP48() { }
                public SubP48(BinaryReader reader) {
                    unknownA = reader.ReadUInt64();
                    unknownB = reader.ReadUInt64();
                }

                public void Write(BinaryWriter writer) {
                    writer.Write(unknownA);
                    writer.Write(unknownB);
                }
            }

            public ulong unknown;
            public List<SubP48> subP48s = new List<SubP48>();

            public P48() { }
            public P48(BinaryReader reader) {
                unknown = reader.ReadUInt64();
                int count = reader.ReadInt32();
                for (int i = 0; i < count; i++) {
                    subP48s.Add(new SubP48(reader));
                }
            }

            public void Write(BinaryWriter writer) {
                writer.Write(unknown);
                writer.Write(subP48s.Count);
                foreach (SubP48 subP48 in subP48s) {
                    subP48.Write(writer);
                }
            }
        }
        public class Pickup {
            public long guid;
            public Item item;

            public Pickup() { }
            public Pickup(BinaryReader reader) {
                guid = reader.ReadInt64();
                item = new Item(reader);
            }

            public void Write(BinaryWriter writer) {
                writer.Write(guid);
                item.Write(writer);
            }
        }
        public class Kill {
            public long killer;
            public long victim;
            public int unknown;
            public int xp;

            public Kill() { }
            public Kill(BinaryReader reader) {
                killer = reader.ReadInt64();
                victim = reader.ReadInt64();
                unknown = reader.ReadInt32();
                xp = reader.ReadInt32();
            }

            public void Write(BinaryWriter writer) {
                writer.Write(killer);
                writer.Write(victim);
                writer.Write(unknown);
                writer.Write(xp);
            }
        }
        public class Damage {
            public ulong target;
            public ulong attacker;
            public float damage;
            public int unknown;

            public Damage() { }
            public Damage(BinaryReader reader) {
                target = reader.ReadUInt64();
                attacker = reader.ReadUInt64();
                damage = reader.ReadSingle();
                unknown = reader.ReadInt32();
            }

            public void Write(BinaryWriter writer) {
                writer.Write(target);
                writer.Write(attacker);
                writer.Write(damage);
                writer.Write(unknown);
            }
        }
        public class Mission {
            public int sectionX;
            public int sectionY;
            public int unknownA;
            public int unknownB;
            public int unknownC;
            public int id;
            public int type;
            public EntityType monsterID;
            public int level;
            public byte unknownE;
            public MissionState state; //0=ready 1=progressing 2=finished
            public short padding;
            public int currentHP;
            public int maxHP;
            public int chunkX;
            public int chunkY;

            public Mission() { }
            public Mission(BinaryReader reader) {
                sectionX = reader.ReadInt32();
                sectionY = reader.ReadInt32();
                unknownA = reader.ReadInt32();
                unknownB = reader.ReadInt32();
                unknownC = reader.ReadInt32();
                id = reader.ReadInt32();
                type = reader.ReadInt32();
                monsterID = (EntityType)reader.ReadInt32();
                level = reader.ReadInt32();
                unknownE = reader.ReadByte();
                state = (MissionState)reader.ReadByte();
                padding = reader.ReadInt16();
                currentHP = reader.ReadInt32();
                maxHP = reader.ReadInt32();
                chunkX = reader.ReadInt32();
                chunkY = reader.ReadInt32();
            }

            public void Write(BinaryWriter writer) {
                writer.Write(sectionX);
                writer.Write(sectionY);
                writer.Write(unknownA);
                writer.Write(unknownB);
                writer.Write(unknownC);
                writer.Write(id);
                writer.Write(type);
                writer.Write((int)monsterID);
                writer.Write(level);
                writer.Write(unknownE);
                writer.Write((byte)state);
                writer.Write(padding);
                writer.Write(currentHP);
                writer.Write(maxHP);
                writer.Write(chunkX);
                writer.Write(chunkY);
            }
        }

        public List<BlockDelta> blockDeltas = new List<BlockDelta>();
        public List<Hit> hits = new List<Hit>();
        public List<Particle> particles = new List<Particle>();
        public List<Sound> sounds = new List<Sound>();
        public List<Shoot> shoots = new List<Shoot>();
        public List<StaticEntity> statics = new List<StaticEntity>();
        public List<ChunkItems> chunkItems = new List<ChunkItems>();
        public List<P48> p48s = new List<P48>();
        public List<Pickup> pickups = new List<Pickup>();
        public List<Kill> kills = new List<Kill>();
        public List<Damage> damages = new List<Damage>();
        public List<PassiveProc> passiveProcs = new List<PassiveProc>();
        public List<Mission> missions = new List<Mission>();

        public ServerUpdate() : base() {
            PacketID = PacketID.ServerUpdate;
        }
        public ServerUpdate(BinaryReader reader) : this() {
            byte[] uncompressed = Zlib.Decompress(reader.ReadBytes(reader.ReadInt32()));
            MemoryStream stream = new MemoryStream(uncompressed);
            BinaryReader r = new BinaryReader(stream);

            var count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                blockDeltas.Add(new BlockDelta(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                hits.Add(new Hit(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                particles.Add(new Particle(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                sounds.Add(new Sound(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                shoots.Add(new Shoot(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                statics.Add(new StaticEntity(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                chunkItems.Add(new ChunkItems(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                p48s.Add(new P48(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                pickups.Add(new Pickup(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                kills.Add(new Kill(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                damages.Add(new Damage(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                passiveProcs.Add(new PassiveProc(r));
            }

            count = r.ReadInt32();
            for (int i = 0; i < count; i++) {
                missions.Add(new Mission(r));
            }
        }

        protected override void WritePacketData(BinaryWriter writer) {
            MemoryStream stream = new MemoryStream();
            BinaryWriter _wr = new BinaryWriter(stream);

            _wr.Write(blockDeltas.Count);
            foreach (BlockDelta blockDelta in blockDeltas) {
                blockDelta.Write(_wr);
            }

            _wr.Write(hits.Count);
            foreach (Hit hit in hits) {
                hit.Write(_wr, false);
            }

            _wr.Write(particles.Count);
            foreach (Particle particle in particles) {
                particle.Write(_wr);
            }

            _wr.Write(sounds.Count);
            foreach (Sound sound in sounds) {
                sound.Write(_wr);
            }

            _wr.Write(shoots.Count);
            foreach (Shoot shoot in shoots) {
                shoot.Write(_wr, false);
            }

            _wr.Write(statics.Count);
            foreach (StaticEntity staticEntity in statics) {
                staticEntity.Write(_wr);
            }

            _wr.Write(chunkItems.Count);
            foreach (ChunkItems chunkItem in chunkItems) {
                chunkItem.Write(_wr);
            }

            _wr.Write(p48s.Count);
            foreach (P48 p48 in p48s) {
                p48.Write(_wr);
            }

            _wr.Write(pickups.Count);
            foreach (Pickup pickup in pickups) {
                pickup.Write(_wr);
            }

            _wr.Write(kills.Count);
            foreach (Kill kill in kills) {
                kill.Write(_wr);
            }

            _wr.Write(damages.Count);
            foreach (Damage damage in damages) {
                damage.Write(_wr);
            }

            _wr.Write(passiveProcs.Count); //npc rClick ??? todo
            foreach (PassiveProc passiveProc in passiveProcs) {
                passiveProc.Write(_wr, false);
            }

            _wr.Write(missions.Count);
            foreach (Mission mission in missions) {
                mission.Write(_wr);
            }

            byte[] data = Zlib.Compress(stream.ToArray());
            stream = new MemoryStream();
            _wr = new BinaryWriter(stream);
            _wr.Write(data.Length);
            _wr.Write(data);

            writer.Write(stream.ToArray());
        }
    }
}

File: src/main/kotlin/packets/ServerUpdate.kt
package com.github.lastexceed.cubeworldnetworking.packets

import io.ktor.util.toByteArray
import io.ktor.utils.io.ByteChannel
import io.ktor.utils.io.close
import com.github.lastexceed.cubeworldnetworking.utils.*

data class ServerUpdate(
	val worldEdits: List<WorldEdit> = emptyList(),
	val hits: List<Hit> = emptyList(),
	val particles: List<Particle> = emptyList(),
	val sounds: List<Sound> = emptyList(),
	val shots: List<Shot> = emptyList(),
	val worldObjects: List<WorldObject> = emptyList(),
	val chunkLoots: List<ChunkLoot> = emptyList(),
	val p48s: List<P48> = emptyList(),
	val pickups: List<Pickup> = emptyList(),
	val kills: List<Kill> = emptyList(),
	val attacks: List<Attack> = emptyList(),
	val statusEffects: List<StatusEffect> = emptyList(),
	val missions: List<Mission> = emptyList()
) : Packet(PacketId.ServerUpdate) {
	override suspend fun writeTo(writer: Writer) {
		val inflatedChannel = ByteChannel(true)
		val inflatedWriter = Writer(inflatedChannel)

		listOf(
			worldEdits,
			hits,
			particles,
			sounds,
			shots,
			worldObjects,
			chunkLoots,
			p48s,
			pickups,
			kills,
			attacks,
			statusEffects,
			missions
		).forEach {
			inflatedWriter.writeInt(it.size)
			it.forEach { subPacket ->
				subPacket.writeTo(inflatedWriter)
			}
		}
		inflatedChannel.close()
		val deflated = Zlib.deflate(inflatedChannel.toByteArray())
		writer.writeInt(deflated.size)
		writer.writeByteArray(deflated)
	}

	companion object : CwDeserializer<ServerUpdate> {
		override suspend fun readFrom(reader: Reader) =
			Reader(Zlib.inflate(reader.readByteArray(reader.readInt()))).run {
				ServerUpdate(
					worldEdits = List(readInt()) { WorldEdit.readFrom(this) },
					hits = List(readInt()) { Hit.readFrom(this) },
					particles = List(readInt()) { Particle.readFrom(this) },
					sounds = List(readInt()) { Sound.readFrom(this) },
					shots = List(readInt()) { Shot.readFrom(this) },
					worldObjects = List(readInt()) { WorldObject.readFrom(this) },
					chunkLoots = List(readInt()) { ChunkLoot.readFrom(this) },
					p48s = List(readInt()) { P48.readFrom(this) },
					pickups = List(readInt()) { Pickup.readFrom(this) },
					kills = List(readInt()) { Kill.readFrom(this) },
					attacks = List(readInt()) { Attack.readFrom(this) },
					statusEffects = List(readInt()) { StatusEffect.readFrom(this) },
					missions = List(readInt()) { Mission.readFrom(this) }
				)
			}
	}
}

data class WorldEdit(
	val position: Vector3<Int>,
	val color: Vector3<Byte>,
	val blockType: BlockType,
	val unknown: Int
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector3Int(position)
		writer.writeVector3Byte(color)
		blockType.writeTo(writer)
		writer.writeInt(unknown)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			WorldEdit(
				position = reader.readVector3Int(),
				color = reader.readVector3Byte(),
				blockType = BlockType.readFrom(reader),
				unknown = reader.readInt()
			)
	}

	enum class BlockType : CwSerializableEnumByte {
		Air,
		Solid,
		Liquid,
		Wet;

		companion object : CwEnumByteDeserializer<BlockType> {
			override val values = values()
		}
	}
}

data class Particle(
	val position: Vector3<Long>,
	val velocity: Vector3<Float>,
	val color: Vector3<Float>,
	val alpha: Float,
	val size: Float,
	val count: Int,
	val type: Type,
	val spread: Float,
	val unknown: Int
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector3Long(position)
		writer.writeVector3Float(velocity)
		writer.writeVector3Float(color)
		writer.writeFloat(alpha)
		writer.writeFloat(size)
		writer.writeInt(count)
		type.writeTo(writer)
		writer.writeFloat(spread)
		writer.writeInt(unknown)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Particle(
				position = reader.readVector3Long(),
				velocity = reader.readVector3Float(),
				color = reader.readVector3Float(),
				alpha = reader.readFloat(),
				size = reader.readFloat(),
				count = reader.readInt(),
				type = Type.readFrom(reader),
				spread = reader.readFloat(),
				unknown = reader.readInt()
			)
	}

	enum class Type : CwSerializableEnumInt {
		Normal,
		Spark,
		Unknown,
		NoSpreadNoRotation,
		NoGravity;

		companion object : CwEnumIntDeserializer<Type> {
			override val values = values()
		}
	}
}

data class Sound(
	val position: Vector3<Float>,
	val type: Type,
	val pitch: Float = 1f,
	val volume: Float = 1f
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector3Float(position)
		type.writeTo(writer)
		writer.writeFloat(pitch)
		writer.writeFloat(volume)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Sound(
				position = reader.readVector3Float(),
				type = Type.readFrom(reader),
				pitch = reader.readFloat(),
				volume = reader.readFloat()
			)
	}

	enum class Type : CwSerializableEnumInt {
		Hit,
		Blade1,
		Blade2,
		LongBlade1,
		LongBlade2,
		Hit1,
		Hit2,
		Punch1,
		Punch2,
		HitArrow,
		HitArrowCritical,
		Smash1,
		SlamGround,
		SmashHit2,
		SmashJump,
		Swing,
		ShieldSwing,
		SwingSlow,
		SwingSlow2,
		ArrowDestroy,
		Blade3,
		Punch3,
		Salvo2,
		SwordHit03,
		Block,
		ShieldSlam,
		Roll,
		Destroy2,
		Cry,
		Levelup2,
		Missioncomplete,
		Watersplash01,
		Step2,
		StepWater,
		StepWater2,
		StepWater3,
		Channel2,
		ChannelHit,
		Fireball,
		FireHit,
		Magic01,
		Watersplash,
		WatersplashHit,
		LichScream,
		Drink2,
		Pickup,
		Disenchant2,
		Upgrade2,
		Swirl,
		HumanVoice01,
		HumanVoice02,
		Gate,
		SpikeTrap,
		FireTrap,
		Lever,
		Charge2,
		Magic02,
		Drop,
		DropCoin,
		DropItem,
		MaleGroan,
		FemaleGroan,
		MaleGroan2,
		FemaleGroan2,
		GoblinMaleGroan,
		GoblinFemaleGroan,
		LizardMaleGroan,
		LizardFemaleGroan,
		DwarfMaleGroan,
		DwarfFemaleGroan,
		OrcMaleGroan,
		OrcFemaleGroan,
		UndeadMaleGroan,
		UndeadFemaleGroan,
		FrogmanMaleGroan,
		FrogmanFemaleGroan,
		MonsterGroan,
		TrollGroan,
		MoleGroan,
		SlimeGroan,
		ZombieGroan,
		Explosion,
		Punch4,
		MenuOpen2,
		MenuClose2,
		MenuSelect,
		MenuTab,
		MenuGrabItem,
		MenuDropItem,
		Craft,
		CraftProc,
		Absorb,
		Manashield,
		Bulwark,
		Bird1,
		Bird2,
		Bird3,
		Cricket1,
		Cricket2,
		Owl1,
		Owl2;

		companion object : CwEnumIntDeserializer<Type> {
			override val values = values()
		}
	}
}

data class WorldObject(
	val chunk: Vector2<Int>,
	val objectID: Int,
	val paddingA: Int,
	val objectType: ObjectType,
	val paddingB: Int,
	val position: Vector3<Long>,
	val orientation: Orientation,
	val size: Vector3<Float>,
	val isClosed: Boolean,
	val transformTime: Int,
	val unknown: Int,
	val paddingC: Int,
	val interactor: Long
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector2Int(chunk)
		writer.writeInt(objectID)
		writer.writeInt(paddingA)
		objectType.writeTo(writer)
		writer.writeInt(paddingB)
		writer.writeVector3Long(position)
		orientation.writeTo(writer)
		writer.writeVector3Float(size)
		writer.writeBoolean(isClosed); writer.pad(3)
		writer.writeInt(transformTime)
		writer.writeInt(unknown)
		writer.writeInt(paddingC)
		writer.writeLong(interactor)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			WorldObject(
				chunk = reader.readVector2Int(),
				objectID = reader.readInt(),
				paddingA = reader.readInt(),
				objectType = ObjectType.readFrom(reader),
				paddingB = reader.readInt(),
				position = reader.readVector3Long(),
				orientation = Orientation.readFrom(reader),
				size = reader.readVector3Float(),
				isClosed = reader.readInt() > 0,
				transformTime = reader.readInt(),
				unknown = reader.readInt(),
				paddingC = reader.readInt(),
				interactor = reader.readLong()
			)
	}

	enum class ObjectType : CwSerializableEnumInt {
		Statue,
		Door,
		BigDoor,
		Window,
		CastleWindow,
		Gate,
		FireTrap,
		SpikeTrap,
		StompTrap,
		Lever,
		Chest,
		ChestTop02,
		Table1,
		Table2,
		Table3,
		Stool1,
		Stool2,
		Stool3,
		Bench,
		Bed,
		BedTable,
		MarketStand1,
		MarketStand2,
		MarketStand3,
		Barrel,
		Crate,
		OpenCrate,
		Sack,
		Shelter,
		Cupboard,
		Desktop,
		Counter,
		Shelf1,
		Shelf2,
		Shelf3,
		CastleShelf1,
		CastleShelf2,
		CastleShelf3,
		StoneShelf1,
		StoneShelf2,
		StoneShelf3,
		SandstoneShelf1,
		SandstoneShelf2,
		SandstoneShelf3,
		Corpse,
		RuneStone,
		Artifact,
		FlowerBox1,
		FlowerBox2,
		FlowerBox3,
		StreetLight,
		FireStreetLight,
		Fence1,
		Fence2,
		Fence3,
		Fence4,
		Vase1,
		Vase2,
		Vase3,
		Vase4,
		Vase5,
		Vase6,
		Vase7,
		Vase8,
		Vase9,
		Campfire,
		Tent,
		BeachUmbrella,
		BeachTowel,
		SleepingMat,
		Furnace,
		Anvil,
		SpinningWheel,
		Loom,
		SawBench,
		Workbench,
		CustomizationBench;

		companion object : CwEnumIntDeserializer<ObjectType> {
			override val values = values()
		}
	}

	enum class Orientation : CwSerializableEnumInt {
		South,
		East,
		North,
		West;

		companion object : CwEnumIntDeserializer<Orientation> {
			override val values = values()
		}
	}
}

data class ChunkLoot(
	val chunk: Vector2<Int>,
	val drops: List<Drop> = listOf()
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector2Int(chunk)
		writer.writeInt(drops.size)
		drops.forEach { it.writeTo(writer) }
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			ChunkLoot(
				chunk = reader.readVector2Int(),
				drops = List(reader.readInt()) { Drop.readFrom(reader) }
			)
	}
}

data class Drop(
	val item: Item,
	val position: Vector3<Long>,
	val rotation: Float,
	val scale: Float,
	val unknownA: Int,
	val droptime: Int,
	val unknownB: Int,
	val unknownC: Int
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		item.writeTo(writer)
		writer.writeVector3Long(position)
		writer.writeFloat(rotation)
		writer.writeFloat(scale)
		writer.writeInt(unknownA)
		writer.writeInt(droptime)
		writer.writeInt(unknownB)
		writer.writeInt(unknownC)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Drop(
				item = Item.readFrom(reader),
				position = reader.readVector3Long(),
				rotation = reader.readFloat(),
				scale = reader.readFloat(),
				unknownA = reader.readInt(),
				droptime = reader.readInt(),
				unknownB = reader.readInt(),
				unknownC = reader.readInt()
			)
	}
}

data class P48(
	val chunk: Vector2<Int>,
	val subPackets: List<ByteArray> = listOf()
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector2Int(chunk)
		writer.writeInt(subPackets.size)
		subPackets.forEach {
			writer.writeByteArray(it)
		}
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			P48(
				chunk = reader.readVector2Int(),
				subPackets = List(reader.readInt()) { reader.readByteArray(16) }
			)
	}
}

data class Pickup(
	val interactor: CreatureId,
	val item: Item
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeLong(interactor.value)
		item.writeTo(writer)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Pickup(
				interactor = CreatureId(reader.readLong()),
				item = Item.readFrom(reader)
			)
	}
}

data class Kill(
	val killer: CreatureId,
	val victim: CreatureId,
	val unknown: Int,
	val xp: Int
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeLong(killer.value)
		writer.writeLong(victim.value)
		writer.writeInt(unknown)
		writer.writeInt(xp)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Kill(
				killer = CreatureId(reader.readLong()),
				victim = CreatureId(reader.readLong()),
				unknown = reader.readInt(),
				xp = reader.readInt()
			)
	}
}

data class Attack(
	val target: Long,
	val attacker: Long,
	val damage: Float,
	val unknown: Int
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeLong(target)
		writer.writeLong(attacker)
		writer.writeFloat(damage)
		writer.writeInt(unknown)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Attack(
				target = reader.readLong(),
				attacker = reader.readLong(),
				damage = reader.readFloat(),
				unknown = reader.readInt()
			)
	}
}

data class Mission(
	val sector: Vector2<Int>,
	val unknownA: Int,
	val unknownB: Int,
	val unknownC: Int,
	val id: Int,
	val type: Int,
	val boss: Race,
	val level: Int,
	val unknownE: Byte,
	val state: State,
	val padding: Short,
	val currentHP: Int,
	val maxHP: Int,
	val chunk: Vector2<Int>
) : SubPacket {
	override suspend fun writeTo(writer: Writer) {
		writer.writeVector2Int(sector)
		writer.writeInt(unknownA)
		writer.writeInt(unknownB)
		writer.writeInt(unknownC)
		writer.writeInt(id)
		writer.writeInt(type)
		boss.writeTo(writer)
		writer.writeInt(level)
		writer.writeByte(unknownE)
		state.writeTo(writer)
		writer.writeShort(padding)
		writer.writeInt(currentHP)
		writer.writeInt(maxHP)
		writer.writeVector2Int(chunk)
	}

	companion object {
		internal suspend fun readFrom(reader: Reader) =
			Mission(
				sector = reader.readVector2Int(),
				unknownA = reader.readInt(),
				unknownB = reader.readInt(),
				unknownC = reader.readInt(),
				id = reader.readInt(),
				type = reader.readInt(),
				boss = Race.readFrom(reader),
				level = reader.readInt(),
				unknownE = reader.readByte(),
				state = State.readFrom(reader),
				padding = reader.readShort(),
				currentHP = reader.readInt(),
				maxHP = reader.readInt(),
				chunk = reader.readVector2Int()
			)
	}

	enum class State : CwSerializableEnumByte {
		Ready,
		InProgress,
		Finished;

		companion object : CwEnumByteDeserializer<State> {
			override val values = values()
		}
	}
}
